teps/0082-workspace-hinting.md

```diff
@@ -0,0 +1,290 @@
+  - name: shared-data
+    default:
+      volumeClaimTemplate:
+        spec:
+          accessModes:
+            - ReadWriteOnce
+          resources:
+            requests:
+              storage: 256M
+```
+
+#### Pros:
+- Very explicit about the Task's expectations for the content provided.
+- Pipelines can override a Tasks' expected types - so, for example, a Task might expect a ConfigMap but a Pipeline might override with a PV instead.
+
+#### Cons:
+- Not just an API change - there will be some logic involved here on the Pipelines controller
+  side to apply the default workspace config to runs.
+- Questions around future extensions stemming from this change are quite nuanced:
+  - What if an author wants the "default" to actually be a requirement and for the TaskRun to fail if it's missing? For example a deploy Task requiring a Secret with name "cluster-key" to exist in the TaskRun's namespace.
+  - What if an author wants to use a default ConfigMap only if that default exists in the TaskRun's namespace but otherwise a fallback like an `emptyDir`?
+  - What if a Catalog Pipeline author attaches a PersistentVolume type or StorageClass that is only available on a subset of cloud providers?
+
+### Add a Hinting Field to Workspace Declarations
+
+> The precise name of this field can be iterated on but for now let's assume "profile".
+
+A Workspace Declaration in a Task or Pipeline can include a `profile` field that is a string matching a fixed set of available options:
+- `"cache"` to hint that the workspace will be used as a cache for performance or reproducibility (e.g. a system might require all teams to use a shared `node_modules` directory when compiling their frontends).
+- `"config"` to hint that the workspace is intended to supply some configuration or settings.
+- `"credential"` to hint that the workspace will be used to perform authenticated actions.
+- `"data"` to hint that the contents are arbitrary data either consumed or produced by the Task.
+
+Each of these profiles can have some fairly reasonable assumptions attached to them. A `"credential"` Workspace likely should be populated from a Secret or Secret-like volume, for example. A `"cache"` Workspace might most often be supplied with a long-lived read-only Persistent Volume. A `"data"` Workspace might be safely assumed to require an ephemeral Persistent Volume that lives only as long as the PipelineRun. Finally `"configuration"` could map consistently to `ConfigMaps`. Most importantly: these specific decisions should be left up to the platform - our own Workflows project may be able to utilize these profiles, for example, to make informed decisions about how to structure a PipelineRun based solely on a Pipeline YAML, supplied list of volumes and set of Secret references.
+
+Here's an example from a `git-clone`-like Task that accepts an optional GitHub deploy key:
+
+```yaml
+workspaces:
+- name: deploy-key
+  readOnly: true
+  optional: true
+  profile: credential
```

Comment:
Does `readOnly` overrides the profile defaults provided by profile `credential` ?

Reply:
`readOnly: true` is added to the `volumeMount` in the step's container (this is how it works today with workspaces). I think that would remain the same after this feature is implemented too.

There aren't any defaults attached to this "credential" profile tho - it's just a string hint to the system creating the PipelineRuns / TaskRuns.

Reply:
Thanks Scott, I was confused by :

>
>Each of these profiles can have some fairly reasonable assumptions attached to them. A `"credential"` Workspace likely should be populated from a Secret or Secret-like volume, for example. A `"cache"` Workspace might most often be supplied with a long-lived read-only Persistent Volume. A `"data"` Workspace might be safely assumed to require an ephemeral Persistent Volume that lives only as long as the PipelineRun.

Reply:
Could this be phrased differently or edited to make it clearer?  I do mention directly after this that `Most importantly: these specific decisions should be left up to the platform` but maybe there's a better way to communicate the point?

Reply:
It might be due to my own limitation. I was also wondering to myself, is this hinting mechanism sufficient to address the said use case i.e. 

>The Tekton Workflows project is currently exploring ways to pass Secrets from a high-level Workflow description into a PipelineRun. This is made considerably more difficult because Pipelines can't indicate which of their Workspaces might be the right one to bind those Secrets to.

e.g. a Pipeline may need multiple secrets from different workspaces, and indicating all of them are `secret` profile -  will it be sufficient information to bind?

Forgive me, if this question is as rookie as the previous.

Reply:
You nailed it: that's the biggest drawback I think. I described it like this in the "Cons" section below:

> Given a Task that optionally accepts both an "ssh-key" and "basic-auth" Workspace marked with the "credential" profile, would an automated system still be able to infer a correct mapping?

I'm going to rewrite this section to clear things up, I reckon it would read better with some edits.

Reply:
Rewritten it, hopefully a bit clearer. Lemme know what you think.

Reply:
Looks good !! 😄